src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2968 @@
+                batch::inner_nofill(tx1),
+                batch::inner(pay(alice, bob, XRP(10)), seq + 2));
+
+            env(jt.jv, batch::sig(bob), ter(temINVALID_INNER_BATCH));
+            env.close();
+        }
+
+        // temINVALID_INNER_BATCH: Batch: inner txn cannot include TxnSignature.
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 2);
+            auto tx1 = pay(alice, bob, XRP(1));
+            tx1[jss::TxnSignature] = "DEADBEEF";
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(tx1, seq + 1),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 2),
+                ter(temINVALID_INNER_BATCH));
+            env.close();
+        }
+
+        // temINVALID_INNER_BATCH: Batch: inner txn must include empty
+        // SigningPubKey.
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 2);
+            auto tx1 = pay(alice, bob, XRP(1));
+            tx1[jss::SigningPubKey] = strHex(alice.pk());
+            tx1[jss::Sequence] = seq + 1;
+            tx1[jss::Fee] = "0";
+            tx1[jss::Flags] = tx1[jss::Flags].asUInt() | tfInnerBatchTxn;
+            auto jt = env.jtnofill(
+                batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner_nofill(tx1),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 2));
+
+            env(jt.jv, ter(temINVALID_INNER_BATCH));
+            env.close();
+        }
+
+        // temINVALID_INNER_BATCH: Batch: inner txn cannot include Signers.
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 2);
+            auto tx1 = pay(alice, bob, XRP(1));
+            tx1[sfSigners.jsonName] = Json::arrayValue;
+            tx1[sfSigners.jsonName][0U][sfSigner.jsonName] = Json::objectValue;
+            tx1[sfSigners.jsonName][0U][sfSigner.jsonName][sfAccount.jsonName] =
+                alice.human();
+            tx1[sfSigners.jsonName][0U][sfSigner.jsonName]
+               [sfSigningPubKey.jsonName] = strHex(alice.pk());
+            tx1[sfSigners.jsonName][0U][sfSigner.jsonName]
+               [sfTxnSignature.jsonName] = "DEADBEEF";
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(tx1, seq + 1),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 2),
+                ter(temINVALID_INNER_BATCH));
+            env.close();
+        }
+
+        // temINVALID_INNER_BATCH: Batch: inner txn preflight failed.
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 2);
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(acctdelete(alice, bob), seq + 1),
+                batch::inner(pay(alice, bob, XRP(-1)), seq + 2),
```

Comment:
A couple of minor notes to make this test a bit easier to read:
```suggestion
                batch::inner(pay(alice, bob, XRP(1)), seq + 1),
                // amount can't be negative
                batch::inner(pay(alice, bob, XRP(-1)), seq + 2),
```

Reply:
fixed: https://github.com/XRPLF/rippled/pull/5060/commits/230f1306f1c130fb221994d1a693dcddfe909e0c